# R/spe_kmers_list.R

#' Title: Function to download the virus genome(s).
#'
#' Description: This function will get the genomes of viruses from NCBI using their accession codes. These accession codes are stored inside a csv file called (Virus_list) that contains a columns with the virus type and subtype as well as one or multiple accession code(s) per virus.
#' This function will then store all viruses genomes in a fasta according to their name. If multiple lines have the same name in the virus_list they will be combined into one fasta file and considered as variation of the virus genome.
#' @param path_csv_file path to the csv file containing a column "name" and a column "accession_code" for viruses of interest.
#' @param path_output path to the directory where the genomes will be stored (one .fasta file will be generated by unique name in the column name of the virus_list).
#' @param virus_name if "all" then all viruses will be downloaded from the virus_list, otherwise the specific virus (1 max) will be downloaded
#' @return Saves the .fasta files inside the "path_output"
#' @export


download_virus <- function(path_csv_file,path_output,virus_name = "all"){
  Virus_csv <- read.csv(path_csv_file)
  if (virus_name == "all"){
    for (name in unique(Virus_csv$name)){
      print(paste("downloading",name))
      ids <- Virus_csv$accession_code[Virus_csv$name == name]
      combined_sequences <- rentrez::entrez_fetch(db = "nuccore", id = ids, rettype = "fasta", retmode = "text")
      write(combined_sequences, file = paste(path_output,name,".fasta",sep = ""))
    }
  }
  else{
    ids <- Virus_csv$accession_code[Virus_csv$name == virus_name]
    print(paste("downloading",name))
    combined_sequences <- rentrez::entrez_fetch(db = "nuccore", id = ids, rettype = "fasta", retmode = "text")
    write(combined_sequences, file = paste(path_output,virus_name,".fasta",sep = ""))
  }
}


download_fastq_vector <- function(accession_id, output_dir, n_reads) {
  dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)
  fastq_dump_command <- c(
    "run",
    "--rm",
    "-v", paste0(normalizePath(output_dir), ":/output"),
    "ncbi/sra-tools",
    "fastq-dump",
    accession_id,
    "--outdir", "/output",
    "--split-files",
    paste0("-X ", n_reads)
  )
  
  # Run the Docker command
  system2("docker", args = fastq_dump_command, stdout = TRUE, stderr = TRUE)
}