---
title: "Download virus and mosquito genomes and RNAseq data"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: html_document
---

Setup of download_virus enabling the download of data from a csv file. This allows to get or all genomes in one command or to select which genome will be downloaded.

```{r}
download_virus <- function(output_path,accession_id){
  combined_sequences <- rentrez::entrez_fetch(db = "nuccore", id = accession_id, rettype = "fasta", retmode = "text")
  write(combined_sequences, file = paste0(output_path,accession_id,".fasta"))
  print(paste0(accession_id," downloaded and stored at --> ",output_path))
}
```

Setup of download function for getting fasq DNAseq data from NCBI with their accession codes. This function need a docker image installation of sra-tools in order to acces the data. The function allows to decide how much sequences for a accession_code will be downloaded allowing to try the pipeline on smaller samples first.

```{r}
download_fastq_vector <- function(accession_id, output_dir, n_reads) {
  dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)
  fastq_dump_command <- c(
    "run",
    "--rm",
    "-v", paste0(normalizePath(output_dir), ":/output"),
    "ncbi/sra-tools",
    "fastq-dump",
    accession_id,
    "--outdir", "/output",
    "--split-files",
    paste0("-X ", n_reads)
  )
  # Run the Docker command
  system2("docker", args = fastq_dump_command, stdout = TRUE, stderr = TRUE)
}
```

# Download Virus genomes.

```{r}
output_path <- "../data_processed/virus/west_nile/"

accession_id <- "LC817237.1"
download_virus(output_path ,accession_id)
```

# Download Vector fastq sequences.

```{r}
accession_id <- "SRR29420355"
output_dir <- "../data_processed/vector/aedes/aegypty/"
number_of_sequences <- "100000"
download_fastq_vector(accession_id, output_dir, number_of_sequences)
```
