```{r}
library(Biostrings)
library(stringr)
library(dplyr)
library(ggplot2)
library(purrr)
library(kmers.profile)
```

```{r}
virus_name <- "dengue_virus_type_1"
k <- 6
x <- 100
size <- "1000000"
```

```{r}
v_fasta_file <- paste("../data/virus/reference/", virus_name, ".fasta", sep = "")
m_fasta_files <- tools::file_path_sans_ext(list.files(path = "../data/vector/chr1/",full.names = FALSE))
#m_fasta_file <- "../data/testing_data/raw_fasta_files/Aedes_aegypti_sequences.fasta"

best_kmers_list <- gen_spe_first_obs_kmers(v_fasta_file,k,x)

v_kmer_pos_df <- gen_kmers_pos_df(v_fasta_file,best_kmers_list,virus_name)
#m_kmer_pos_df <- gen_kmers_pos_df(m_fasta_file,best_kmers_list,m_fasta_files)

kmer_pos_list <- list()
for (m_fasta_file in m_fasta_files) {
  m_kmer_pos_df <- gen_kmers_pos_df(paste("../data/vector/chr1/", m_fasta_file, ".fna", sep = ""), best_kmers_list, m_fasta_file)
  kmer_pos_list[[m_fasta_file]] <- m_kmer_pos_df
  write.csv(m_kmer_pos_df, file = paste("../data/kmers/", m_fasta_file, "_kmer_pos_df.csv", sep = ""), row.names = FALSE)}
combined_kmer_pos_df <- do.call(rbind, kmer_pos_list)

# GENERATE REFERENCE VALUES

prob <- 1 / (4^k) 
geom_values <- rgeom(100, prob)
log_geom_values <- log10(geom_values + 1)

# GET T_TEST DF

results_df <- get_t_test(v_kmer_pos_df,m_kmer_pos_df)
```

```{r}
# PLOT REGUARDING KMER OF INTEREST

kmer = "GATGGA"
geom_data <- data.frame(Kmer = "Reference_distribution", 
                         log10_segm_size = log_geom_values)
plot_data <- bind_rows(
  v_kmer_pos_df %>% 
    filter(Kmer == kmer) %>%
    mutate(Source = "Virus"),
  m_kmer_pos_df %>% 
    filter(Kmer == kmer) %>%
    mutate(Source = "Mosquito"),
  geom_data %>% mutate(Source = "Reference_distribution")
)


means_df <- plot_data %>%
  group_by(Source) %>%
  summarise(mean_value = log10(mean(10^(log10_segm_size), na.rm = TRUE)), .groups = 'drop')

line_value <- log10(4^k)
custom_colors <- c("Virus" = "cadetblue1", 
                   "Mosquito" = "darkseagreen2", 
                   "Reference_distribution" = "gray")

na_positions <- plot_data %>%
  filter(is.na(log10_segm_size), Kmer == kmer, Source == "Virus") %>%
  select(Source, Position)

plot <- ggplot(plot_data, aes(x = Source, y = log10_segm_size, fill = Source, color = Source)) +  
  geom_violin(trim = FALSE, scale = "width", alpha = 0.5) +  
  geom_jitter(size = 1, width = 0.2, alpha = 0.3) +  
  geom_point(data = means_df, aes(x = Source, y = mean_value), color = "black", size = 2, shape = 21, fill = "black") +  
  geom_hline(yintercept = line_value, linetype = "dashed", color = "coral") +  
  labs(title = paste("Segment Sizes for Kmer chosen for WN: ", kmer, sep = ""), y = "Log10 Segment Size", x = "") +  
  theme_minimal() +
  ylim(0.8,4.5)+
  theme(legend.position = "none",  
        plot.title = element_text(size = 15, face = "bold"),  
        axis.text.x = element_text(size = 15, angle = 35, hjust = 1),  # Rotate X-axis labels by 30Â°
        axis.text.y = element_text(size = 10),  
        axis.title.y = element_text(size = 10),  
        panel.background = element_rect(fill = "white"),
        plot.background = element_rect(fill = "white")) +  
  scale_fill_manual(values = custom_colors) +  
  scale_color_manual(values = custom_colors) +  
  geom_text(aes(x = 2.5, y = line_value, label = paste("Expected size using a random ",k,"-mer", sep = "")),  
            color = "coral", vjust = -0.5, size = 5) +  
  geom_jitter(data = na_positions, aes(x = Source, y = log10(Position)), color = "blue", size = 1, width = 0.2, alpha = 0.4, height = 0)

plot
ggsave("../result/___violin_plot_spe_west_nile_small.png", plot, width = 5, height = 5, dpi = 300)
```

```{r}
virus_name <- "dengue_virus_type_1"
k <- 6
x <- 100
size <- "1000000"
```

```{r}
v_fasta_file <- paste("../data/virus/reference/", virus_name, ".fasta", sep = "")

best_kmers_list <- gen_spe_first_obs_kmers(v_fasta_file,k,x)

v_kmer_pos_df <- gen_kmers_pos_df(v_fasta_file,best_kmers_list,virus_name)


get_t_test_single <- function(df_virus) {  
  virus_values <- df_virus %>%
    group_by(Kmer) %>%
    summarize(log10_segm_size = list(na.omit(log10_segm_size)), .groups = 'drop')
  results <- data.frame(Kmer = character(),
                        mean_virus = numeric(),
                        sd_virus = numeric(),
                        stringsAsFactors = FALSE)
  for (kmer in virus_values$Kmer) {
    virus_sizes <- unlist(virus_values$log10_segm_size[virus_values$Kmer == kmer])
    mean_virus <- ifelse(length(virus_sizes) > 0, mean(virus_sizes, na.rm = TRUE), NA)
    sd_virus <- ifelse(length(virus_sizes) > 1, sd(virus_sizes, na.rm = TRUE), NA)
    results <- rbind(results, data.frame(
      Kmer = kmer,
      mean_virus = mean_virus,
      sd_virus = sd_virus
    ))
  }
  return(results)
}

results <- get_t_test_single(v_kmer_pos_df)

```


```{r}
count_specific_kmer_in_fastq <- function(fq, specific_kmer) { 
  sequences <- as.character(sread(fq)) 
  kmer_count <- 0 
  total_size <- 0 
  for (seq in sequences) { 
    total_size <- total_size + nchar(seq)
    kmer_count <- kmer_count + (gregexpr(specific_kmer, seq, fixed = TRUE)[[1]] != -1) %>% sum() 
  } 
  ratio <- if (total_size > 0) { 
    kmer_count / total_size 
  } else { 
    0 
  } 
  return(list(kmer_count = kmer_count, ratio = ratio))
}
```

```{r}
calculate_kmer_frequencies <- function(results_df, fq) {
  kmers <- results_df$Kmer
  
  results_df$kmer_count <- numeric(length(kmers))
  results_df$ratio <- numeric(length(kmers))
  
  for (i in seq_along(kmers)) {
    kmer <- kmers[i]
    result <- count_specific_kmer_in_fastq(fq, kmer)  
    results_df$kmer_count[i] <- result$kmer_count     
    results_df$ratio[i] <- result$ratio                 
  }
  
  return(results_df)
}

fq <- readFastq("../data/testing_data/SRR23079314_1.fastq")
results_df <- calculate_kmer_frequencies(results, fq)

results_df$frequency_virus <- 1 / (10 ^ results_df$mean_virus)
```

```{r}
library(ggplot2)

# Assuming results_df already contains the necessary columns
# Create new columns for the X and Y axes
results_df$x_value <- 1 / results_df$sd_virus  # X-axis values
results_df$y_value <- results_df$frequency_virus / results_df$ratio  # Y-axis values

# Plotting with kmer names as labels
ggplot(results_df, aes(x = x_value, y = y_value)) +
  geom_text(aes(label = Kmer), vjust = -0.5, size = 3) +  # Using geom_text to label with Kmer names
  labs(title = "Scatter Plot of 1/sd_virus vs log2(frequency_virus/ratio)",
       x = "1/sd_virus",
       y = "log2(frequency_virus/ratio)") +
  theme_minimal() +
  theme(axis.text.x = element_text(hjust = 1)) 

```



